### PR TITLE
Duplication of content and ingest items

### PR DESCRIPTION
**Pull request by marwoodandrew**
[SD-1290] As a User, I want the ability to fetch previously fetched content.
[SD-815] As a user, I want to duplicate content items

Major changes

* Fetch action allowed from both ingest and archive.
* Fetch is allowed more than once from ingest.
* New guids are created with each ingest.
* Original guids are saved in new field ingest_id
* New field family_id allows related items to be identified.
* All versions are duplicated along with content items (deep copy)

### CODE REVIEW

File: apps/archive/archive.py [hunk 1]
```diff
@@ -265,6 +266,48 @@ def restore_version(self, id, doc):
         doc.update(old)
         return res
 
+    def duplicate_content(self, original_doc):
+        if original_doc.get('type', '') == 'composite':
```
Comment (petrjasek) [inline apps/archive/archive.py]: is the idea to duplicate also contents of package?

File: apps/archive/archive.py [hunk 2]
```diff
@@ -265,6 +266,48 @@ def restore_version(self, id, doc):
...
+
+        new_versions = []
+        for old_version in old_versions:
+            old_version['_id_document'] = new_doc['_id']
+            del old_version['_id']
+            old_version['guid'] = new_doc['guid']
+            old_version['unique_name'] = new_doc['unique_name']
+            old_version['unique_id'] = new_doc['unique_id']
+            old_version['versioncreated'] = utcnow()
+            new_versions.append(old_version)
```
Comment (amagdas) [inline apps/archive/archive.py]: Why is this needed, using `old_versions` should do the trick...

File: apps/archive/archive_ingest.py [hunk 3]
```diff
@@ -23,11 +23,15 @@
...
 from superdesk.utc import utcnow
 from superdesk.resource import Resource
+from .common import generate_guid, generate_unique_id_and_name, GUID_TAG
 from superdesk.services import BaseService
 from .archive import SOURCE as ARCHIVE
 from superdesk.workflow import is_workflow_state_transition_valid
 from apps.content import LINKED_IN_PACKAGES, PACKAGE
+from superdesk.notification import push_notification
 STATE_FETCHED = 'fetched'
+FAMILY_ID = 'family_id'
```
Comment (amagdas) [inline apps/archive/archive_ingest.py]: I think the constants should be in `app.content.py` so we can use them without depending on `archive`

File: apps/archive/archive_ingest.py [hunk 4]
```diff
@@ -43,25 +47,40 @@ class ArchiveIngestResource(Resource):
...
         for doc in docs:
             ingest_doc = superdesk.get_resource_service('ingest').find_one(req=None, _id=doc.get('guid'))
             if not ingest_doc:
-                msg = 'Fail to found ingest item with guid: %s' % doc.get('guid')
-                raise SuperdeskApiError.notFoundError(msg)
-
-            if not is_workflow_state_transition_valid('fetch_as_from_ingest', ingest_doc[config.CONTENT_STATE]):
-                raise InvalidStateTransitionError()
+                # see if it is in archive, if it is duplicate it
+                archived_doc = superdesk.get_resource_service(ARCHIVE).find_one(req=None, _id=doc.get('guid'))
```
Comment (petrjasek) [inline apps/archive/archive_ingest.py]: would there be a document in archive with `_id` equal to ingest item `guid`? because it seems that archive doc `_id` is generated to a new guid. maybe searching for `family_id` would do the trick. but in general the item in ingest should be there so it might just fail if it can't be found

File: apps/archive/archive.py (hunk 2 above)
Comment (petrjasek) [inline apps/archive/archive.py]: `old_versions` reference old document, not the new copy of it

Comment (amagdas) [inline apps/archive/archive.py]: No copy is being done, the old doc is being updated...

File: apps/archive/archive.py (hunk 1 above)
Comment (akintolga) [inline apps/archive/archive.py]: Yes, one requirement was to do a deep copy of the packages. That is to create duplicates of every item in a package. Except the renditions.

File: apps/archive/archive.py (hunk 2 above)
Comment (akintolga) [inline apps/archive/archive.py]: I was thinking similar to Petr but I'll double check that one tomorrow again.

Comment (petrjasek) [inline apps/archive/archive.py]: on post it should create new items for each item without an `_id` so this should create copies

File: apps/archive/archive_ingest.py (hunk 4 above)
Comment (marwoodandrew) [inline apps/archive/archive_ingest.py]: What we're trying to do is to work out if we are fetching from ingest or fetching from the archive (duplicating). So assuming that the guid passed is the is found in ingest it's a fetch from ingest, if it is not we check if the guid is in the archive, that is a duplication. We always generate a new guid in fetch.

Comment (petrjasek) [inline apps/archive/archive_ingest.py]: ok didn't realize it's a single endpoint for both

File: apps/archive/archive.py (hunk 2 above)
Comment (amagdas) [inline apps/archive/archive.py]: `_id` is deleted, so using them directly should be ok...

Comment (petrjasek) [inline apps/archive/archive.py]: I see, you're right

Comment (akintolga) [inline apps/archive/archive.py]: I think the reason I created a new list was this: old_versions is a cursor and once it is eventuated (iterated) it couldn't be iterated again. So if you use old_document in the post you'll get errors.

Comment (amagdas) [inline apps/archive/archive.py]: It makes a lot of sense, sorry for this :blush: 
:+1:

**Comment by amagdas** [comment]
:+1:

File: apps/archive/archive.py (hunk 2 above)
Comment (akintolga) [inline apps/archive/archive.py]: No problem Adrian. Actually thanks for highlighting this area. I spotted one potential issue: I think I need to check -before posting- if the list has any items in it. So still room for improvement :)

Comment (marwoodandrew) [inline apps/archive/archive.py]: Added check

**Comment by amagdas** [comment]
:+1:

**Comment by petrjasek** [comment]
:+1:
